crates/admin/src/cluster_controller/service/state.rs

```diff
@@ -277,80 +281,779 @@ where
         Ok(())
     }
 
+    #[instrument(
+        level = "debug",
+        skip(self),
+        fields(
+            partition_table_version = ?self.partition_table_watcher,
+            logs_metadata_version = ?self.cluster_state_watcher.current().logs_metadata_version
+        ),
+    )]
     async fn trim_logs(&self) {
-        let result = self.trim_logs_inner().await;
+        let cluster_state = self.cluster_state_watcher.current();
+        let current_trim_points = match get_current_trim_points(&cluster_state, &self.bifrost).await
+        {
+            Ok(trim_points) => trim_points,
+            Err(err) => {
+                warn!("Failed to get current logs trim points. This can lead to increased log server disk usage: {err}");
+                return;
+            }
+        };
+
+        let new_trim_points = TrimMode::from(self.snapshots_repository_configured, cluster_state)
+            .safe_trim_points(&current_trim_points, Some(self.log_trim_threshold));
+        debug!(?new_trim_points, "Safe trim points");
 
-        if let Err(err) = result {
-            warn!("Could not trim the logs. This can lead to increased disk usage: {err}");
+        for (log_id, (trim_point, partition_id)) in new_trim_points {
+            info!(
+                %partition_id,
+                "Trimming log {log_id} records ..='{trim_point}'"
+            );
+            let result = self.bifrost.admin().trim(log_id, trim_point).await;
+
+            if let Err(err) = result {
+                warn!(
+                    %partition_id,
+                    "Failed to trim log {log_id}. This can lead to increased log server disk usage: {err}"
+                );
+            }
         }
     }
+}
 
-    async fn trim_logs_inner(&self) -> Result<(), restate_bifrost::Error> {
-        let cluster_state = self.cluster_state_watcher.current();
+async fn get_current_trim_points(
+    cluster_state: &ClusterState,
+    bifrost: &Bifrost,
+) -> Result<HashMap<LogId, Lsn>, restate_bifrost::Error> {
+    let partition_ids: Vec<PartitionId> = cluster_state
+        .nodes
+        .values()
+        .filter_map(|node_state| match node_state {
+            NodeState::Alive(alive_node) => Some(alive_node),
+            _ => None,
+        })
+        .flat_map(|node| node.partitions.keys())
+        .cloned()
+        .collect();
```

Comment:
While I hope we remove the need to do this get_trim_point for every log (hint, this can be concurrent instead of sequential). Here are a few tips to improve this code a little:
1- No need to allocate the Vec<> since you consume the iterator immediately after
2- You can use `.copied()` instead of `.cloned()` to make it easier for the reader to know that those are copy types.

Here is another take:
```rust
let partition_ids = cluster_state
    .nodes
    .values()
    .filter_map(|node_state| match node_state {
        NodeState::Alive(alive_node) => Some(alive_node.partitions.keys()),
        _ => None,
    })
    .flatten();
 for partition in partition_ids {
    let log_id = LogId::from(*partition);
    ...
 }
```